File: hw/toapanta/u3/hw35CrudStrategy/CrudStrategyJS/models/strategies/CsvStrategy.js
import fs from 'fs/promises';
import path from 'path';
import { Customer } from '../Customer.js';
import { CrudStrategy } from './CrudStrategy.js';

export class CsvStrategy extends CrudStrategy {
    constructor() {
        super();
        this.filePath = path.join(process.cwd(), 'data', 'customers.csv');
        this.ensureFileExists();
    }

    async ensureFileExists() {
        try {
            const dirPath = path.dirname(this.filePath);
            await fs.mkdir(dirPath, { recursive: true });
            
            try {
                await fs.access(this.filePath);
            } catch {
                await fs.writeFile(this.filePath, '');
            }
        } catch (error) {
            console.error(`Error ensuring CSV file exists: ${error.message}`);
        }
    }

    async add(customer) {
        try {
            const customers = await this.readAll();
            
            const existingCustomer = customers.find(c => c.id === customer.id);
            if (existingCustomer) {
                console.log(`Error: ID ${customer.id} already exists in CSV`);
                return false;
            }
            
            const csvLine = customer.toCsv();
            await fs.appendFile(this.filePath, csvLine + '\n');
            
            console.log(`Customer added successfully to CSV. ID: ${customer.id}`);
            return true;
            
        } catch (error) {
            console.error(`Error adding customer to CSV: ${error.message}`);
            return false;
        }
    }

    async delete(customerId) {
        try {
            const customers = await this.readAll();
            const initialCount = customers.length;
            
            const filteredCustomers = customers.filter(c => c.id !== customerId);
            
            if (filteredCustomers.length < initialCount) {
                const content = filteredCustomers.map(c => c.toCsv()).join('\n');
                await fs.writeFile(this.filePath, content);
                
                console.log(`Customer ${customerId} deleted from CSV. Remaining: ${filteredCustomers.length}`);
                return true;
            }
            
            console.log(`Customer ${customerId} not found in CSV`);
            return false;
            
        } catch (error) {
            console.error(`Error deleting from CSV: ${error.message}`);
            return false;
        }
    }

    async update(customerId, customer) {
        try {
            const customers = await this.readAll();
            let updated = false;
            
            for (let i = 0; i < customers.length; i++) {
                if (customers[i].id === customerId) {
                    customer.id = customerId; 
                    customers[i] = customer;
                    updated = true;
                    break;
                }
            }
            
            if (updated) {
                const content = customers.map(c => c.toCsv()).join('\n');
                await fs.writeFile(this.filePath, content);
                
                console.log(`Customer ${customerId} updated in CSV`);
                return true;
            }
            
            console.log(`Customer ${customerId} not found for update in CSV`);
            return false;
            
        } catch (error) {
            console.error(`Error updating in CSV: ${error.message}`);
            return false;
        }
    }

    async readAll() {
        try {
            const data = await fs.readFile(this.filePath, 'utf8');
            if (!data || data.trim() === '') {
                return [];
            }
            
            const lines = data.trim().split('\n');
            const customers = [];
            
            for (const line of lines) {
                if (line.trim()) {
                    const customer = Customer.fromCsv(line);
                    if (customer) {
                        customers.push(customer);
                    }
                }
            }
            
            console.log(`Loaded ${customers.length} customers from CSV`);
            return customers;
            
        } catch (error) {
            console.error(`Error reading CSV: ${error.message}`);
            return [];
        }
    }

    async readById(customerId) {
        const customers = await this.readAll();
        return customers.find(c => c.id === customerId) || null;
    }

    getFormatName() {
        return "CSV";
    }
}

File: hw/aymacana/u3/hw35CrudStrategy/CrudStrategyPY/model/strategy/csv_strategy.py
import os
from typing import List
from model.customer import Customer
from model.strategy.crud_strategy import CrudStrategy
from utils.file_manager import FileManager

class CsvStrategy(CrudStrategy):
    FILE_PATH = "customers.csv"
    
    def add(self, customer: Customer) -> bool:
        try:
            existing_customers = self.read_all()
            for c in existing_customers:
                if c.id == customer.id:
                    print(f"Error: ID {customer.id} already exists in CSV")
                    return False
            
            csv_line = customer.to_csv()
            FileManager.append_to_file(self.FILE_PATH, csv_line)
            return True
            
        except Exception as e:
            print(f"Error adding customer to CSV: {e}")
            return False
    
    def delete(self, customer_id: int) -> bool:
        try:
            customers = self.read_all()
            initial_count = len(customers)
            
            customers = [c for c in customers if c.id != customer_id]
            
            if len(customers) < initial_count:
                content = "\n".join([c.to_csv() for c in customers])
                FileManager.save_to_file(self.FILE_PATH, content)
                return True
            
            return False
            
        except Exception as e:
            print(f"Error deleting from CSV: {e}")
            return False
    
    def update(self, customer_id: int, customer: Customer) -> bool:
        try:
            customers = self.read_all()
            updated = False
            
            for i, c in enumerate(customers):
                if c.id == customer_id:
                    customer.id = customer_id  
                    customers[i] = customer
                    updated = True
                    break
            
            if updated:
                content = "\n".join([c.to_csv() for c in customers])
                FileManager.save_to_file(self.FILE_PATH, content)
            
            return updated
            
        except Exception as e:
            print(f"Error updating in CSV: {e}")
            return False
    
    def read_all(self) -> List[Customer]:
        try:
            content = FileManager.read_file(self.FILE_PATH)
            if not content:
                return []
            
            lines = content.strip().split("\n")
            customers = []
            
            for line in lines:
                if line.strip():  
                    customer = Customer.from_csv(line)
                    if customer:
                        customers.append(customer)
            
            return customers
            
        except Exception as e:
            print(f"Error reading CSV: {e}")
            return []
    
    def read_by_id(self, customer_id: int) -> Customer:
        customers = self.read_all()
        for customer in customers:
            if customer.id == customer_id:
                return customer
        return None
    
    def get_format_name(self) -> str:
        return "CSV"